luavm/src/lib/lua_values/mod.rs

```diff
@@ -57,7 +57,16 @@ impl LuaVal {
                 // The behaviour of `as f64` is the same as the conversion
                 // from int to float described in the manual.
                 LuaValKind::INT => Ok(((self.val >> tagging::TAG_SHIFT) as i64) as f64),
-                LuaValKind::FLOAT => Ok(transmute::<usize, f64>(self.val ^ self.kind() as usize)),
+                LuaValKind::FLOAT => {
+                    // in the case of floats, the lower bits contain the tag, thus
+                    // before returning the actual value, the bits should be cleared
+                    // if any of the low order bits are set, then the implementation
+                    // boxes the float. This ensures that the tag does not overwrite
+                    // the bits of the float
+                    Ok(transmute::<usize, f64>(
+                        self.val ^ LuaValKind::FLOAT as usize,
```

Comment:
I'm not personally sure what the precedence of `^` and `as` are (I assume `^` binds tighter, but I don't know). It might be worth using brackets to make this clearer?